### ISSUE TITLE
Add blacklist/whitelist of signal and synchronous update types

### ISSUE THREAD

**Issue by mfateev** [temporalio/temporal#4387]
**Is your feature request related to a problem? Please describe.**
Signals, even ones that workflow doesn't care about, still consume workflow history. In some use cases, signals are expected only in certain workflow states, and signals received in other states should be rejected.

**Describe the solution you'd like**
Workflow should support blacklisting and whitelisting signal and update types it cares about. Blacklisted signals and updates are rejected immediately without updating the mutable state and the history.

**Comment by lorensr** [comment]
Implementation: At Workflow Execution start time, you can specify an allow/denylist. List is mutable by command from Workflow.

In future, if we have a way for Workers to tell Server per-WorkflowType configuration, a default allow/denylist could be colocated with the Workflow definition, which in many cases would be better DX IMO.

**Comment by danthegoodman1** [comment]
And update on this? We've had to move a lot _off_ temporal because of this

**Comment by danthegoodman1** [comment]
Ideally we'd want to just be able to dynamically change an allow-list

**Comment by lorensr** [comment]
@danthegoodman1 You can currently implement an Update handler (in all SDKs but PHP and TS, which is [coming soon](https://github.com/temporalio/sdk-typescript/pull/1277)) that rejects any Updates that aren't in your (dynamic) allowlist. See Validation phase: https://docs.temporal.io/workflows#update

**Comment by danthegoodman1** [comment]
Yeah problem with that is mixing timeouts and such becomes hacky, but it’s a workaround
